woocommerce/facebook-for-woocommerce
# Clean up background sync options

Closes #1943
Closes #1948

~~**_Note_ this PR is based off #1953 `feature/add-cron-heartbeat` branch** because it uses the new hourly heartbeat event to schedule the cleanup task. I plan on merging this to master after #1953 is merged (`blocked` label).~~

### Changes proposed in this Pull Request:
This PR adds a cleanup task to remove `wp_options` rows for old, completed background sync jobs. 

As noted in the issue, the [SkyVerge plugin framework](https://github.com/skyverge/wc-plugin-framework) has a system for managing background jobs - see `SV_WP_Background_Job_Handler `. Each job is represented by a `wp_options` row, and they are not cleaned up.

Facebook for WooCommerce is using this to run a [background sync](https://github.com/woocommerce/facebook-for-woocommerce/blob/master/includes/Products/Sync/Background.php#L31) regularly. This is a periodic task, so is causing more and more options to accumulate over time (which could cause performance issues).

It's possible that this system is intended for one-off tasks. The other tasks that use it (`Background_Remove_Duplicate_Visibility_Meta`, `Background_Handle_Virtual_Products_Variations`) look like one-off jobs to me (based on comments), and I don't see any options rows for them in my dev site's database.

This PR only cleans up product sync jobs - `wc_facebook_background_product_sync_job_`. If other jobs are accumulating options or causing issues for merchants, we can clean those up too (in a separate PR/follow up).

### How to test the changes in this Pull Request:
1. Investigate to confirm your test site's database has redundant option rows - e.g. `SELECT * FROM `wp_options` WHERE option_name like 'wc_facebook_background%' `.
2. Switch to this branch. Wait 1+ days (e.g. leave overnight).
3. Confirm that the number of rows reduces over time (job cleans up incrementally, a few at a time). Check that the oldest ones are deleted first.
4. ~~View Facebook log in `WooCommerce > Status > Logs` and confirm there's a clear message logging how many rows were deleted.~~ (logging removed in 2cac5f656761bfd8bc76eb2261d640f951c0f3d7)

Also please confirm that this doesn't break or impact product sync, or delete options that it shouldn't.

### Changelog entry

> Fix - Clean up old product sync background job `wp_options` rows #1962
